### ISSUE TITLE
Support for decorators / custom babel config

### ISSUE THREAD

**Issue by lukethacoder** [cyco130/detype#7]
Is there a way to add support for experimental decorators?

Have added `experimentalDecorators: true` and `emitDecoratorMetadata: true` to the `tsconfig.json` but still getting errors. Would allowing the consumer to provide a custom babel config bypass user specific issues like this?

```
SyntaxError: C:\Users\PROJECT_PATH\testFile.ts: Support for the experimental syntax 'decorators' isn't currently enabled (12:3):

  10 |    * @param {string | number}
  11 |    */
> 12 |   @api value;
     |   ^
  13 | /* @detype: empty-line=2 */
  14 |   /**
  15 |    * @param {string | number}
```

**Comment by cyco130** [comment]
Hi,

I just published `0.6.3` that adds the `customizeBabelConfig` option. It will call the function you provide there with the babel config as an escape hatch for customizing it. I think this should solve your problem.

**Comment by lukethacoder** [comment]
Whilst this silences the issues, it ends up mangling the output, which defeats the purpose of this package.

I've got an [example repo](https://github.com/lukethacoder/detype-lwc-ts-example/tree/main) of what I'm trying to achieve. More just need the transformer/compiler to ignore the decorators, not actually convert them to anything.

This is my test `.ts` input

```ts
import { LightningElement, api } from "lwc";

interface TestInterfaceHere {
  label: string;
  value?: string | number | object;
}

export default class TypescriptLwc extends LightningElement {
  /**
   * @param {string | number}
   */
  @api value;

  /**
   * @param {string | number}
   */
  @api label;

  get labelAndValue() {
    return {
      value: this.value,
      label: this.label
    } as TestInterfaceHere;
  }
}
```


And this is ideally what I'd be looking for in the output `*.js`:

```js
import { LightningElement, api } from "lwc";

export default class TypescriptLwc extends LightningElement {
  /**
   * @param {string | number}
   */
  @api value;

  /**
   * @param {string | number}
   */
  @api label;

  get labelAndValue() {
    return {
      value: this.value,
      label: this.label
    };
  }
}
```

but instead am getting this `*.js` output:

```js
var _class, _descriptor, _descriptor2;
function _initializerDefineProperty(target, property, descriptor, context) {
  if (!descriptor) return;
  Object.defineProperty(target, property, {
    enumerable: descriptor.enumerable,
    configurable: descriptor.configurable,
    writable: descriptor.writable,
    value: descriptor.initializer
      ? descriptor.initializer.call(context)
      : void 0,
  });
}
function _defineProperty(obj, key, value) {
  key = _toPropertyKey(key);
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true,
    });
  } else {
    obj[key] = value;
  }
  return obj;
}
function _toPropertyKey(arg) {
  var key = _toPrimitive(arg, "string");
  return typeof key === "symbol" ? key : String(key);
}
function _toPrimitive(input, hint) {
  if (typeof input !== "object" || input === null) return input;
  var prim = input[Symbol.toPrimitive];
  if (prim !== undefined) {
    var res = prim.call(input, hint || "default");
    if (typeof res !== "object") return res;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return (hint === "string" ? String : Number)(input);
}
function _applyDecoratedDescriptor(
  target,
  property,
  decorators,
  descriptor,
  context
) {
  var desc = {};
  Object.keys(descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;
  if ("value" in desc || desc.initializer) {
    desc.writable = true;
  }
  desc = decorators
    .slice()
    .reverse()
    .reduce(function (desc, decorator) {
      return decorator(target, property, desc) || desc;
    }, desc);
  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }
  if (desc.initializer === void 0) {
    Object.defineProperty(target, property, desc);
    desc = null;
  }
  return desc;
}
function _initializerWarningHelper(descriptor, context) {
  throw new Error(
    "Decorating class property failed. Please ensure that " +
      "proposal-class-properties is enabled and runs after the decorators transform."
  );
}
import { LightningElement, api } from "lwc";

let TypescriptLwc =
  ((_class = class TypescriptLwc extends LightningElement {
    constructor(...args) {
      super(...args);
      /**
       * @param {string | number}
       */ _initializerDefineProperty(this, "value", _descriptor, this);

      /**
       * @param {string | number}
       */ _initializerDefineProperty(this, "label", _descriptor2, this);
    }

    get labelAndValue() {
      return {
        value: this.value,
        label: this.label,
      };
    }
  }),
  ((_descriptor = _applyDecoratedDescriptor(_class.prototype, "value", [api], {
    configurable: true,
    enumerable: true,
    writable: true,
    initializer: null,
  })),
  (_descriptor2 = _applyDecoratedDescriptor(_class.prototype, "label", [api], {
    configurable: true,
    enumerable: true,
    writable: true,
    initializer: null,
  }))),
  _class);
export { TypescriptLwc as default };

```

**Comment by lukethacoder** [comment]
Bit more poking around I realised I only needed the `syntax` version of the babel plugin, not the parse and transform plugin.

Thanks for opening up the custom babel config. Managed to get it working with the below for anyone else looking to use decorators with this package.

```json
{
  "plugins": [
    ["@babel/plugin-syntax-decorators", { "version": "legacy" }]
  ]
}
```

[@babel/plugin-syntax-decorators](https://babeljs.io/docs/babel-plugin-syntax-decorators)
